### PR TITLE
Expose new .netcoreapp1.1 APIs in System.Runtime

### LINKED ISSUE #692: Add a non-generic version of Enum.TryParse
These are the methods in Enum class related to string parsing.
```c#
public static object Parse(Type enumType, string value);
public static object Parse(Type enumType, string value, bool ignoreCase);

public static bool TryParse<TEnum>(string value, out TEnum result) 
where TEnum : struct;
public static bool TryParse<TEnum>(string value, bool ignoreCase, out TEnum result) 
where TEnum : struct;
```
Apparently Parse is exposed only in non-generic form, while TryParse is only exposed in generic form. This is somewhat inconsistent.

It is simple enough to call a non-generic method in a generic method, but not the other way around. So Parse can be made generic using the following helper. But it is a lot harder to wrap TryParse<T> into TryParse(Type enumType) without reflection calls.
```c#
public static TEnum Parse<TEnum>(string value, bool ignoreCase)
{
    return Enum.Parse(typeof(TEnum), value, ignoreCase);
}
```
The suggestion is to add a non-generic Enum.TryParse method, the API signature would look like
```c#
public static bool TryParse(Type enumType, string value, out object result) 
public static bool TryParse(Type enumType, string value, bool ignoreCase, out object result) 
```

### LINKED ISSUE #2352: API Addition: Add generic Array.Reverse<T>
Add generic `Array.Reverse<T>`.

## Rationale

[`Array.Reverse`](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/mscorlib/src/System/Array.cs#L1570-L1606) is non-generic. It includes a fast path for a known set of primitive types via a call into the runtime (see [`TrySZReverse`](https://github.com/dotnet/coreclr/blob/4cf8a6b082d9bb1789facd996d8265d3908757b2/src/classlibnative/bcltype/arrayhelpers.cpp#L390-L424)), otherwise it falls back to slower code paths that involve boxing for value types. This is a significant performance cost for non-primitive value types, ~22x slower.

Callers of `Array.Reverse` (like `List<T>.Reverse` and `ImmutableArray<T>.Builder.Reverse`) are affected by this performance issue. As a workaround, [`List<T>.Reverse`](https://github.com/dotnet/coreclr/pull/1231) and [`ImmutableArray<T>.Builder.Reverse`](https://github.com/dotnet/corefx/pull/2354) are being updated to not use the non-generic `Array.Reverse` method until the generic `Array.Reverse<T>` method becomes available.

Other non-generic methods on `Array` have generic counterparts like `Array.Sort` and `Array.Sort<T>`, so this would just be making `Reverse` consistent with those.

## Proposed API

```c#
public abstract class Array : ...
{
    // Proposed
    public static void Reverse<T>(T[] array);
    public static void Reverse<T>(T[] array, int index, int length);

    // Existing
    public static void Reverse(Array array);
    public static void Reverse(Array array, int index, int length);
}
```

### LINKED ISSUE #8592: IntPtr and UIntPtr should implement the IEquatable<T> interface
Both are structs, overload ==, and even have [specialized implementations](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs#L64-L67) for the default `EqualityComparer`, yet they don't implement `IEquatable<T>` (or override `Equals`). Should we consider making them implement the interface?

If there's no particular reason why they shouldn't, then here is the API proposal:

```cs
public struct IntPtr : IEquatable<IntPtr>
{
    public bool Equals(IntPtr other);
}

public struct UIntPtr : IEquatable<UIntPtr>
{
    public bool Equals(UIntPtr other);
}
```

### LINKED ISSUE #10127: Add new Lazy<T>(T value) ctor to System.Runtime contract and add tests
Now that the implementation of #2578 has been merged for CoreCLR (https://github.com/dotnet/coreclr/pull/6304) and CoreRT (https://github.com/dotnet/corert/pull/1538), the new API needs to be exposed in the [`System.Runtime`](https://github.com/dotnet/corefx/blob/4fccde2c42f7670ebd098513c09b30772f833e84/src/System.Runtime/ref/System.Runtime.cs#L1202) contract (and versioned as appropriate) and tests added.

### LINKED ISSUE #11090: Make RuntimeHelpers.TryEnsureSufficientExecutionStack public
This method is very handy to detect upcoming stack overflows in recursive algorithms without the need for exception handling logic. Note that `TrySetFromTask` in https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.Threading.Tasks/src/System/Threading/Tasks/TaskExtensions.CoreCLR.cs has a remark on the possibility for this method to become public. Another place where this would have been useful is in the suggested fix for #7953 .

### PR DESCRIPTION
**Pull request by hughbe**
And add some tests

Fixes #10127
Fixes #8592
Fixes #2352
Fixes #692
Fixes #11090

/cc @justinvp @jamesqo @jkotas @yufeih @bartdesmet @weshaggard @danmosemsft

### CODE REVIEW

**Comment by hughbe** [comment]
@yufeih I think there might be a bug in your implementation of the non-generic `TryParse`. An exception is thrown if `enumType` is null or is not an enum. I personally wouldn't expect `TryParse` to ever throw an exception, and to just return `false` in these cases.

**Comment by yufeih** [comment]
I followed the behavior of *TryParse<TEnum>*

The following code throws ArgumentException:

~~~
int a;
Enum.TryParse<int>("--", out a);
~~~

**Comment by hughbe** [comment]
Ah okay, that's fine then. Was just checking!

**Comment by hughbe** [comment]
@dotnet-bot test this please

File: src/System.Runtime/tests/System/EnumTests.cs [hunk 1]
```diff
@@ -11,24 +11,29 @@ namespace System.Tests
...
+            yield return new object[] { "Blue,Red,Green", false, SimpleEnum.Red | SimpleEnum.Blue | SimpleEnum.Green };
+            yield return new object[] { "Blue,Red,Red,Red,Green", false, SimpleEnum.Red | SimpleEnum.Blue | SimpleEnum.Green };
+            yield return new object[] { "Red,Blue,   Green", false, SimpleEnum.Red | SimpleEnum.Blue | SimpleEnum.Green };
+            yield return new object[] { "1", false, SimpleEnum.Red };
+            yield return new object[] { " 1 ", false, SimpleEnum.Red };
+            yield return new object[] { "2", false, SimpleEnum.Blue };
+            yield return new object[] { "99", false, (SimpleEnum)99 };
+            yield return new object[] { "-42", false, (SimpleEnum)(-42) };
+            yield return new object[] { "   -42", false, (SimpleEnum)(-42) };
+            yield return new object[] { "   -42 ", false, (SimpleEnum)(-42) };
```
Comment (weshaggard) [inline src/System.Runtime/tests/System/EnumTests.cs]: Is there a reason to prefer this approach over the InlineData approach? I don't have a strong preference I'm just trying to understand the motivation

File: src/System.Runtime/tests/System.Runtime.Tests.csproj [hunk 2]
```diff
@@ -28,6 +28,7 @@
     <Compile Include="Helpers.cs" />
     <Compile Include="System\ActivatorTests.cs" />
     <Compile Include="System\ArrayTests.cs" />
+    <Compile Include="System\ArrayTests.netcoreapp1.1.cs" Condition="'$(TargetGroup)' == 'netcoreapp1.1'" />
```
Comment (weshaggard) [inline src/System.Runtime/tests/System.Runtime.Tests.csproj]: I would prefer these all being grouped in one ItemGroup with that ItemGroup having the condition.

File: src/System.Runtime/tests/System/EnumTests.cs (hunk 1 above)
Comment (weshaggard) [inline src/System.Runtime/tests/System/EnumTests.cs]: Nevermind I see now you did this so you can share the test cases.

**Comment by weshaggard** [comment]
Looks like some of the System.Runtime tests are hitting some kind of interesting infinite loop or something in Release builds for your new set of tests. Please try to take a look at a Release test run for your new configuration.

File: src/System.Runtime/tests/System.Runtime.Tests.csproj (hunk 2 above)
Comment (hughbe) [inline src/System.Runtime/tests/System.Runtime.Tests.csproj]: done, thanks

**Comment by hughbe** [comment]
Thanks Wes, turns out it was a bad test - the release CLR inlined a method, meaning that it never reached stack limit. Fixed by adding a `NoInlining` attribute

File: src/System.Runtime/tests/project.json [hunk 3]
```diff
@@ -32,6 +32,11 @@
       "dependencies": {
         "System.Runtime.Serialization.Formatters": "4.3.0-beta-24530-03"
       }
+    },
+    "netcoreapp1.1": {
+      "dependencies": {
+        "System.Runtime.Serialization.Formatters": "4.3.0-beta-24530-03"
```
Comment (weshaggard) [inline src/System.Runtime/tests/project.json]: What test actually has a dependency on this library?

Comment (hughbe) [inline src/System.Runtime/tests/project.json]: I'm afk right now so can't remove the dependency and see what fails, but it was introduced in https://github.com/dotnet/corefx/commit/f6f97c206fa6b9edc48e1335cdc3d3d33cb54590#diff-29df68f37e76e464ad7ef9130076abae

Comment (weshaggard) [inline src/System.Runtime/tests/project.json]: Thanks for the pointer looks like @alexperovich added it to do some BinaryFormatter testing. I'm OK with this for testing purposes.

File: src/System.Runtime/tests/System/Runtime/CompilerServices/RuntimeHelpersTests.netcoreapp1.1.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        [Fact]
+        public static void TryEnsureSufficientExecutionStack_NoSpaceAvaiable_ReturnsFalse()
+        {
+            // This test will fail with a StackOverflowException if TryEnsureSufficientExecutionStack() doesn't
+            // return false. No exception is thrown and the test finishes when TryEnsureSufficientExecutionStack()
+            // returns true.
+            if (!RuntimeHelpers.TryEnsureSufficientExecutionStack())
+            {
+                Assert.Throws<InsufficientExecutionStackException>(() => RuntimeHelpers.EnsureSufficientExecutionStack());
+                return;
```
Comment (hughbe) [inline src/System.Runtime/tests/System/Runtime/CompilerServices/RuntimeHelpersTests.netcoreapp1.1.cs]: Yes. If I remove this check, a StackOverflow exception is thrown - the lines below your comment recurse the method

**Pull request merged**